### PR TITLE
Attempt to prevent CME due to off-thread block entity updates

### LINKED ISSUE #366: CME in Rendering Data Attachment
There is a CME on this line: [MixinChunkRendererRegion.java#L45](https://github.com/FabricMC/fabric/blob/1.14/fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java#L45)

This was first reported on the optifabric issue tracker [here](https://github.com/modmuss50/OptiFabric/issues/37) but it has since been reproduced without optifine. (I think optifine is just making it more common).

[crash-2019-09-12_22.24.59-client.txt](https://github.com/FabricMC/fabric/files/3610773/crash-2019-09-12_22.24.59-client.txt)

As for a way to reproduce it without optifine im not sure of one. One possible idea would be the world lagging and it happening when it was trying to catch up? (suggestion from Fudge on discord)

### PR DESCRIPTION
**Pull request by grondag**
Attempt to fix #366 

Test build is [here](https://github.com/grondag/fabric/releases/download/Fix366Test1/fabric-rendering-data-attachment-v1-0.1.1+1436934f42.jar).  

The crash is intermittent and doesn't happen for everyone, so this should not be merged until we get some field reports to indicate that it works.

### CODE REVIEW

File: fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java [hunk 1]
```diff
@@ -16,52 +16,132 @@
...
+				// CMEs when we iterate the map.  (Vanilla does not iterate these maps when it builds
+				// the chunk cache and does not suffer from this problem.)
+				// 
+				// We handle this by trying to iterating the block entity map first, because that 
+				// usually works and is most efficient.  If we encounter a CME, we retry, querying the 
+				// map by iterating block positions. This method is a little slower but will not cause a CME.
+				try {
+					map = mapChunkFast(chunk, posFrom, posTo, map);
+				} catch (ConcurrentModificationException e) {
+					map = mapChunkSafely(chunk, posFrom, posTo, map);
```
Comment (liach) [inline fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java]: consider logging this cme for future debug purposes
you can also remove the map argument

Comment (grondag) [reply fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java]: Not understanding what you mean be remove the map argument. it is needed in all cases for this to function properly.

Comment (liach) [reply fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java]: no... you always pass in a null for the map. just remove that param

**Comment by grondag** [comment]
The more I consider this, the more I think the entire module should be deprecated and scrapped in 1.15. The deprecated implementation could simply forward to the per-chunk block-entity hash maps. In 1.15, models that need block entity state would obtain it via the block view they can already access.

The intention of this module - to ensure thread safety and consistency of block entity state access - cannot be achieved in a modded environment without adding our own synchronization mechanisms to the underlying vanilla data structures.  IMO that is too invasive and prone to causing other problems as bad or worse than the current CME crash.

Thread safety isn't strictly required for models to look up block entity state.  The hash maps might return inconsistent (or null) values but a simple lookup shouldn't crash if the event of concurrent modification. For example, if the block entity state has been modified or deleted during or after the creation of the render cache, then the model may get state that is more recent than the cached block state.

This means models must take basic precautions (handle nulls, make sure the block entity and block state match, as examples) that more or less make this module superfluous. 

Note that the functional impact of such events is typically minor - inconsistency typically  implies the chunk will need to be rebuilt again almost immediately or is now unloaded, so any visual inconsistency should be short-lived if seen at all.

File: fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java (hunk 1 above)
Comment (grondag) [reply fabric-rendering-data-attachment-v1/src/main/java/net/fabricmc/fabric/mixin/rendering/data/client/MixinChunkRendererRegion.java]: map is only guaranteed to be null on the first iteration - note that it is inside a loop

**Comment by grondag** [comment]
Per my comments earlier, I've deprecated this entire module.  The implementation is a pass-through to the vanilla block entity lookup logic, and mods that rely on block entity state during rendering must take precautions when doing so during chunk rebuilds. (Which was already the case.)

**Comment by grondag** [comment]
After more reviewing, I still don't trust this as-is.  Too many scenarios where `getBlockEntity` could be problematic if called off-thread.  I could revert the deprecation, but that was barely better. Need to consider this some more.

For now, either test build should help players experiencing crashes.

**Comment by grondag** [comment]
Alright, reverting deprecation until we have something better.  I have some notions but they seem too extensive.  This should at least be much less likely to break. This can be reviewed now.

**Pull request merged**